src/main/java/com/o19s/es/explore/ExplorerQueryBuilder.java

```diff
@@ -0,0 +1,143 @@
+import org.elasticsearch.common.io.stream.NamedWriteable;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.index.query.AbstractQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryParseContext;
+import org.elasticsearch.index.query.QueryShardContext;
+
+import java.io.IOException;
+import java.util.Objects;
+import java.util.Optional;
+
+public class ExplorerQueryBuilder extends AbstractQueryBuilder<ExplorerQueryBuilder> implements NamedWriteable{
+    public static final String NAME = "match_explorer";
+
+    private static final ParseField QUERY_NAME = new ParseField("query");
+    private static final ParseField TYPE_NAME = new ParseField("type");
+
+    private QueryBuilder query;
+    private String type;
+
+    private static final ObjectParser<ExplorerQueryBuilder, QueryParseContext> PARSER;
+
+    static {
+        PARSER = new ObjectParser<>(NAME, ExplorerQueryBuilder::new);
+        PARSER.declareObject(
+                ExplorerQueryBuilder::query,
+                (parser, context) -> context.parseInnerQueryBuilder().get(),
+                QUERY_NAME
+        );
+        PARSER.declareString(ExplorerQueryBuilder::statsType, TYPE_NAME);
+        declareStandardFields(PARSER);
+    }
+
+
+    public ExplorerQueryBuilder() {
+    }
+
+    public ExplorerQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        query = in.readNamedWriteable(QueryBuilder.class);
+        type = in.readString();
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeNamedWriteable(query);
+        out.writeString(type);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+
+        printBoostAndQueryName(builder);
+
+        builder.field(QUERY_NAME.getPreferredName());
+        query.toXContent(builder, params);
+
+        builder.field(TYPE_NAME.getPreferredName(), type);
+
+        builder.endObject();
+    }
+
+    public static Optional<ExplorerQueryBuilder> fromXContent(QueryParseContext context) throws IOException {
+        XContentParser parser = context.parser();
+        final ExplorerQueryBuilder builder;
+
+        try {
+            builder = PARSER.parse(context.parser(), context);
+        } catch (IllegalArgumentException iae) {
+            throw new ParsingException(parser.getTokenLocation(), iae.getMessage(), iae);
+        }
+
+        if (builder.query == null) {
+            throw new ParsingException(parser.getTokenLocation(), "Field [" + QUERY_NAME + "] is mandatory.");
+        }
+        if (builder.statsType() == null) {
```

Comment:
for later: when we'll have n enum it'll be easy to check that stats type is invalid and throw a ParsingException here.